### PR TITLE
Cppclass defining improvements

### PR DESCRIPTION
**Pull request by samaingw**
# What it this about

This patch is about fixing and extending some stuff in the C++ support of Cython.
Basically the goal is to improve code definition within a C++ class, to be able to write C++ classes entirely in Cython, not only wrap them.

This patch comes from my current internship at Nexedi, where my goal is to extend Cython to have the possibility to do OOP in a efficient multithreaded context with memory management.
As we can't rely on Python class for that purpose (because of the GIL in CPython), we've choosen the path of C++ class to do the current implementation.
The memory management of this new type of class is done by an atomic refcount, so it's really basic, but, still, it frees the user from manual memory (de)allocation.

This patch is not about this memory managed class. It's a collection of changes I have made while wandering in the C++ class Cython code, trying to make my new type of class work.
So this patch is only about the cppclass system that exists within Cython.
I expect the only blocker would be technical misunderstanding from my part, which lead me to write things that shouldn't have been written.

I will release (hopefully) soon another patch (based on this one), which is the extension of Cython with the above-mentioned memory managed class.
You may reject it for technical reasons, as well as for some sort of "we just don't want to maintain/integrate this into mainline Cython", which would be perfectly acceptable.

This patch contains four parts sorted like the description order below.

The first batch is quite messy as it contains a mix between "perceived non-controversial changes" (because it's not feature but pure bugfix, and small ones) and small features necessary for the rest above to work.

1.a is the most "perceived controversial" commit among them for me, that's why it's the first one: to make you shout if you're unhappy with it.

## 1 - Various small fixes (a useful and crystal-clear name right ?)
a) Be able to call `__init__` and `__dealloc__` in user code
b) Allow docstrings for cppclass (syntactic sugar, it is discarded like comments during code generation)
c) Make CppClassScope holds parent_type like CClassScope
d) Fix is_subclass in the case of an empty base_class attribute
e) Fix void return types for `__init__` and `__dealloc__`
f) Fix some entry attributes: `is_cfunction` and `func_cname`
g) Make `InPlaceAssignmentNode` effectively produce a `BinopNode` with the inplace flag set

## 2 - Overloaded methods
The goal here is to be able to define (and not just declare) methods in CppClass that can be overloaded.
I will now take some time to explain why the current implementation cannot do this, and why I think the approach proposed in my commits is a not-so-bad one.

TL;DR: The current code doesn't allow clean method inheritance.

In the current implementation of `CClassScope`, when we declare a cfunction, we check for the presence
of an entry with the same name, and if this is the case, we shout if the function signature is different or if the entry is already defined.
So we cannot have declaration and/or definition of different function types with the same name.
You can declare then define a function, as it's the same function signature, so everything is fine.

In the current implementation of `CppClassScope`, when we declare a cfunction, we check for the presence of an entry with the same name and if we try to define, not just declare, the new entry.
If this is the case, we update the type if the signatures (function types) are compatible, else we shout.
In any other cases (no previous entry or new entry is just declared, not defined), override the entry.
This checks are performed in `CppClassScope.declare_var`. This also explains why the cpp methods are in the `var_entries` array.
Polymorphism is usable because in `CppClassScope.declare_cfunction` (which calls the above-mentioned `declare_var`), we link the previously declared entries to the new one if we're just declaring the entry, not defining it.
Moreover, in this `declare_cfunction`, we never touch `cfunc_entries`. This explains why this array is always empty in a `CppClassScope`.
For a classic (non-special) method, the flow is:
CppClassScope.declare_cfunction (will link the alternatives to have polymorphism)
----> calls CppClassScope.declare_var (make the checks, if we create a new entry, add it to var_entries)
-----------> calls Scope.declare (blindly create and override the entry in the entries dict)

Wrapping C++ class is possible, because we fall into the "override the entry" clause (we never define an entry).

A real issue is method inheritance.
When inheriting methods with the same name, only one is kept because of the continuous override in `Scope.declare`.
The kept method will be, in my current understanding of the code, the last (in the base) upper (not deep in the inheritance chain: the "parenthesied" bases, not base of base) right-most (in the parenthesis) base method with this name:
```
cdef cppclass Base1:
    int method(int a)

cdef cppclass Base2:
    int method(double a)

cdef cppclass HiddenBase:
    int method(double b, double b)

cdef cppclass Base3(HiddenBase):
    int method(double b, int a)
    int method(int a, double b) # <=== This one will be the only method kept in Derived by inheritance

cdef cppclass Derived(Base1, Base2, Base3):
    pass
    # The only method Derived is aware of is Base3.method(int, double), the other ones were (sadly) lost in the process
```
So we may have lost a lot of alternatives.
This is exactly what is described in issue #1357.

The proposed approach deals with the alternative management directly in `Scope.declare`.
The behaviour is quite simple: we scan the previous alternatives for a compatible function signature.
If there is a match and the previous method is marked as inherited (or the default init), we override it.
If there is a match and it's not an inherited method, shout.
If there is no match, add the new entry to the alternatives.

This involves six changes:
a) Change `Symtab.declare` logic for cppclass methods: the `overloaded_entry` array is properly managed in this code (the switch from is_cpp to is_cpp_class_scope is intentional: it restricts this overloading to cppclass, otherwise it is available to everything when cython produces c++ code)
b) Change the `declare_var` and `declare_cfunction` in CppClassScope to use the previous changes
c) Add the `overloaded_alternative` array to ubcm_entry created with a Type.method() statement
d) Amend a test case (cpp_no_constructor) because the error message is not the same:
Below is a long text explaining why this is not a stupid idea to modify this test.
Beware, there is e) and f) change sets under it.

TL;DR: Method alternatives order in the entries corresponds to the declaration order. Before the patch, the alternatives order corresponded to reversed declaration order.

`PyrexTypes.best_match` shouts when it does not find a suitable alternative.
When doing so, it uses the "principal alternative" as a reference.
Principal alternative here refers to the entry which is in `entries[name]` in the type scope.
The other ones are held in the `overloaded_alternatives` array of this principal alternative.

Before the a) change set, alternatives in a `CppClassScope` are always overridden, and it is the duty
of `CppClassScope.declare_cfunction` to fill the `overloaded_alternative` array of the newly created alternative.
So the principal alternative is always changing, and the last alternative declared will effectively be the principal one.
Consequently, when `PyrexTypes.best_match` shouts, it uses the last declared alternative as a reference (for argument number and such).

After the a and b change set, `Scope.declare` takes care of the alternatives, and overrides an alternative in a `CppClassScope` only if it is an inherited one with the same signature as the new one. As the signature is the same, each info used by `PyrexTypes.best_match` when shouting doesn't change between old overridden alternative and new one.
Consequently, a signature represented by an alternative never change position: if it is carried by the principal alternative, it will always be carried by the principal one, therefore `PyrexTypes.best_match`, when taking the principal alternative as a reference, is effectively taking the first defined alternative as reference.

As `cpp_no_constructor` is an error test about failing to find the correct alternative, it expects `PyrexTypes.best_match` to shout.
The expected string refers to the last (2nd / 2 defined alternatives) alternative, whereas with a & b changes, it should refer to the first alternative, because that's what is expected.

This commit does this.

e) Amend a test case (cpp_classes_def) because one warning must stand:
It is, again, a long text explaining why I'm not mad.

TL;DR: The test module lacked a _WARNINGS variable to catch the RegularPolygon.area warning since the beginning. It's just that it wasn't noticeable before.

In `cpp_classes_def` (in the tests/run directory), we have some defined method (area) in a class (RegularPolygon) which has no exception indication, but is performing a division.
Consequently, a divide-by-zero check is performed, and we should, theorically, raise an exception if the check is true.
As there is no except indication on this method, we have a compile-time warning : Unraisable exception in function 'RegularPolygon.area'.
This is not a bug, it's a feature.

So we should have a warning from cython when compiling this test module. A warning that should be reported when running the test suite.
With the a) & b) changes in Symtab, this is what's happening.
This wasn't before, and the reason is quite tricky.

`RegularPolygon` has a base class, `Shape`, which is declared as 'extern'. Shape has the `area` method.
So, RegularPolygon inherits area from Shape before overloading it.
`Shape.area` and `RegularPolygon.area` have exactly the same signature.
In `CppClassScope.declare_var`, we are in the following situation: previous entry and defining.
This leads to:
*We update the type if the signatures (function types) are compatible, else we shout*
As types are compatible, we just modify the entry type, keeping the inherited entry for all other attributes.
The side effect is that we didn't change `entry.pos`. An inherited entry has a pos equal to None, because the corresponding attribute is declared nowhere in the subclass body.

Consequently, when making the warning, the entry has no pos therefore it is discarded by the test suite when trying to match actual warnings from expected ones.

This commit adds the warning from RegularPolygon.area to the expected warnings. It also removes werror to the tags, since this warning is not a runtime blocker.

f) Provide test for the feature (which is failing on the current master as it is related to issue #1357)

## 3 - Cppclass generation fixes
a) Allow defined (not wrapped) cppclass to have multiple constructors
This fixes cppclass generation. With the above code everything is there to handle overloaded constructors.
This is about generating the correct C++ code when we're not wrapping an external cppclass, but defining it in cython.

## 4 - Nested classes
This is not about wrapping nested C++ classes, because it is already present and functional (and there's a test which can prove it).
This is about being able to generate C++ classes definitions with nested classes. So it is exactly like the batch about method overloading: being able to define and not just declare.
If I have understood it properly, this addresses issue #1218

a) Make cppclass generate code for nested definitions (for example nested cppclass)
b) Fix Symtab for the inherited nested types: this is basically the same thing than for other attributes, have a separate "inherited types" array, to be able to have them in scope but ignore them during code generation.
c) Amend the nested classes test

### CODE REVIEW

File: Cython/Compiler/ExprNodes.py [hunk 1]
```diff
@@ -6949,6 +6949,8 @@ def analyse_as_type_attribute(self, env):
                         ubcm_entry.func_cname = entry.func_cname
                         ubcm_entry.is_unbound_cmethod = 1
                         ubcm_entry.scope = entry.scope
+                        if type.is_cpp_class:
+                            ubcm_entry.overloaded_alternatives = [Symtab.Entry(entry.name, entry.func_cname, entry.type) for entry in entry.overloaded_alternatives]
```
Comment (scoder) [inline Cython/Compiler/ExprNodes.py]: An `Entry` is actually never created like this, rather through some `declare_*()` method in a Scope (sub-)class. I'm not sure right now if this qualifies as a special case. Why do we need a copy here?

File: Cython/Compiler/Parsing.py [hunk 2]
```diff
@@ -3782,6 +3782,7 @@ def p_cpp_class_definition(s, pos,  ctx):
         s.next()
         s.expect('NEWLINE')
         s.expect_indent()
+        doc = p_doc_string(s)
```
Comment (scoder) [inline Cython/Compiler/Parsing.py]: This doesn't seem to be used. Maybe not assign it at all and add a comment instead that it's optional and useless?

Comment (samaingw) [reply Cython/Compiler/Parsing.py]: You're absolutely right, this will be clearer.

File: Cython/Compiler/ExprNodes.py (hunk 1 above)
Comment (samaingw) [reply Cython/Compiler/ExprNodes.py]: Stating the obvious: I'm filling the `entry.overloaded_alternatives` array because polymorphism is allowed in a cppclass, and PyrexTypes.best_match must be able to choose among the alternatives.

Please notice the ubcm_entry creation code is `ubcm_entry = Symtab.Entry(entry.name, cname, ctype)` **but**, in the case of a cppclass, `cname = entry.func_cname` (with the exception of fused_types, a cfunction entry in CppClassScope always has a func_cname).

After setting things in ubcm_entry we will return it, wrapped in a NameNode which will uses `ubcm_entry.cname` during code generation.

`ubcm_entry.cname = entry.func_cname` is then required to have a full namespace path when generating the C++ code. This way we will have `Namespace::Class::Subclass::method` in the generated code. With `ubcm_entry.cname = entry.cname` we would just have `method` in the generated code, and the C++ compiler would be misleaded. This is the same problem for the alternatives in ubcm_entry.overloaded_alternatives.

We cannot call a declare_*() for the alternatives, because, like ubcm_entry, we are not creating a new entry, we are just copying it (with a cname change).

**Comment by jpfeuffer** [comment]
Hi! What is the status of this PR or its split-up PRs? This looks like an incredibly helpful addition and it looks like a lot of work was put in already.
What is missing for example?

**Comment by da-woods** [comment]
@jpfeuffer Out of interest, what's your use-case for defining nested C++ classes? As far as I know the main two things the "nestedness" gives you in C++ are
1. keeping the class out of the main namespace
2. controlling access to it with `private`/`protected`

We definitely don't support the latter in Cython so I'm not quite completely sure what the benefit is.

(With that said, I think it'd still be nice to try to use some of this work - just trying to understand what it lets you do)

**Comment by samaingw** [comment]
After a long absence, I'm coming back to put a bit of work in the following months for theses sub-PRs.
@jpfeuffer I'm glad someone finds this useful !

What remains on my part for now is the review of @da-woods about the tests of #3221. I'll need a bit of time to look back to my code to understand it again. I hope I'll be able to land it during the next two weeks or so.

@da-woods : PR #3220 contains unit test for the `cppclass` docstrings that were in #3183 which is now merged. Could you take some time to look into #3220 ? That's a short PR, just a dumb test.

I guess the beast named #3235 will be for later ^^.

**Comment by jpfeuffer** [comment]
Cool! Our use case is indeed 

> 1. keeping the class out of the main namespace

Currently, we monkey patch the resulting python classes and attach "non-namespaced" classes (which we prefix with "_" to hide them). And we hoped there would be something better, something built-in.
